### PR TITLE
Use cloud-controller-manager by default on CI

### PR DESCRIPTION
**Pull request by aojea**
/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
/test pull-kubernetes-e2e-gce

the test ran, but some failures, let this soak

/test

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -2207,6 +2207,8 @@ function start-kube-controller-manager {
   fi
   if [[ -n "${RUN_CONTROLLERS:-}" ]]; then
     params+=("--controllers=${RUN_CONTROLLERS}")
+  elif [[ "${CLOUD_PROVIDER_FLAG:-external}" == "external" ]]; then
+    params+=("--controllers=*,-nodeipam")
```
Comment (andrewsykim) [inline cluster/gce/gci/configure-helper.sh:2211]: TODO: remove if we merge https://github.com/kubernetes/kubernetes/pull/114596

Comment (aojea) [reply cluster/gce/gci/configure-helper.sh:2211]: this PR is WIP, I prefer to get the fixes in before moving it in tree

**Comment by aojea** [comment]
/cc @liggitt 

the idea is to use a pinned version of the cloud-provider-gcp controller manager as default in the CI, WDYT?

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -2207,6 +2207,8 @@ function start-kube-controller-manager {
   fi
   if [[ -n "${RUN_CONTROLLERS:-}" ]]; then
     params+=("--controllers=${RUN_CONTROLLERS}")
+  elif [[ "${CLOUD_PROVIDER_FLAG:-external}" == "external" ]]; then
```
Comment (liggitt) [inline cluster/gce/gci/configure-helper.sh:2210]: this interaction between RUN_CONTROLLERS and CLOUD_PROVIDER_FLAG isn't obvious... should we always tack on -nodeipam if `"${CLOUD_PROVIDER_FLAG:-external}" == "external"`, whether RUN_CONTROLLERS is set or not?

**Comment by liggitt** [comment]
> the idea is to use a pinned version of the cloud-provider-gcp controller manager as default in the CI, WDYT?

does that mean the code in the nodeipam controller in k/k is untested in presubmit by default?

**Comment by andrewsykim** [comment]
> does that mean the code in the nodeipam controller in k/k is untested in presubmit by default? (e.g. it would have to merge, get pushed to the controller repo, get synced to cloud-provider-gcp, get built and tagged there, then get bumped here before we would notice the code didn't work properly?)

Yes -- not just nodeipam, but also service, route and node controller. Testing the latest changes in k/k against a real provider implementation should be done in the respective repos (k8s.io/cloud-provider-gcp). The primary goal of running GCP controller manager in k/k is to ensure we can provision fully functioning clusters when in-tree providers are  removed. 

In the future, we may have a job in k8s.io/cloud-provider-gcp that periodically pulls in latest k/k so we find regressions faster (@cheftako has a "last known goo" doc that covers this in more details).

**Comment by liggitt** [comment]
> Testing the latest changes in k/k against a real provider implementation should be done in the respective repos (k8s.io/cloud-provider-gcp)

Hmm... as long as the code lives in k/k, tests have to be able to run on that code prior to merge, right?

**Comment by andrewsykim** [comment]
> Testing the latest changes in k/k against a real provider implementation should be done in the respective repos (k8s.io/cloud-provider-gcp)

Sorry -- I should clarify that this only refers to code changes in the `k8s.io/cloud-provider` module that would not be tested in presubmit, because we are installing the GCP provider that is using some N-1 version. 

> Hmm... as long as the code lives in k/k, tests have to be able to run on that code prior to merge, right?

Unit/integration tests, yes. But for e2e tests, we would install a known good version of GCP cloud-controller-manager that is compatible against the code being tested in the PR.

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (cheftako) [reply cluster/gce/gci/configure-helper.sh:2210]: I think the default answer is yes. 

There is a special case if you are migrating from KCM only to KCM + CCM, where the answer might be no. If you want to migrate one controller at a time (which the code supports), there would be a period where you disable nodeipam on the CCM and don't disable it on the KCM. However that is not the normal behavior. 

Generally if you are running the CCM (CLOUD_PROVIDER_FLAG == "external" and your running on GCP) then you want to disable the nodeipam controller on the KCM.

**Comment by cheftako** [comment]
> In the future, we may have a job in k8s.io/cloud-provider-gcp that periodically pulls in latest k/k so we find regressions faster (Walter Fender has a "last known good" doc that covers this in more details).

Actually DangerOnTheRanger's [doc](https://docs.google.com/document/d/1U1YRLsAAZsVs6VyrXj8hkfA6aJp9nDZsG5NsfZ1POPE/edit?resourcekey=0-iVuzWZn3mxL2Qil-4peFFQ#heading=h.fxpk50cps4zs).

File: cluster/gce/gci/configure-helper.sh (hunk 1 above)
Comment (aojea) [reply cluster/gce/gci/configure-helper.sh:2211]: should I remove this now, since the PR merged?

**Comment by aojea** [comment]
@liggitt @andrewsykim @cheftako can we discuss this to come to a conclusion before kep freeze?

doing a change like this once people start to merge features is very risky

**Comment by aojea** [comment]
good news on the testing area, we are able to test cloud-provider-gcp/master build from kubernetes/master against kubernetes@latest-ci https://github.com/kubernetes/cloud-provider-gcp/pull/482

See https://testgrid.k8s.io/provider-gcp-presubmits#pull-cloud-provider-gcp-scenario-kops-simple

File: cluster/gce/config-default.sh [hunk 3]
```diff
@@ -550,12 +545,12 @@ export TLS_CIPHER_SUITES=""
...
-export CLOUD_PROVIDER_FLAG="${CLOUD_PROVIDER_FLAG:-gce}"
+export CLOUD_PROVIDER_FLAG="${CLOUD_PROVIDER_FLAG:-external}"
 
 # When ENABLE_AUTH_PROVIDER_GCP is set, following flags for out-of-tree credential provider for GCP
 # are presented to kubelet:
 # --image-credential-provider-config=${path-to-config}
 # --image-credential-provider-bin-dir=${path-to-auth-provider-binary}
 # Also, it is required that DisableKubeletCloudCredentialProviders and KubeletCredentialProviders
 # feature gates are set to true for kubelet to use external credential provider.
-ENABLE_AUTH_PROVIDER_GCP="${ENABLE_AUTH_PROVIDER_GCP:-false}"
```
Comment (aojea) [inline cluster/gce/config-default.sh:561]: should I modify this feature gates mentioned in the comment too?

`DisableKubeletCloudCredentialProviders and KubeletCredentialProviders`

Comment (liggitt) [reply cluster/gce/config-default.sh:561]: I guess I would have expected ENABLE_AUTH_PROVIDER_GCP=true to drive setting those feature gates?

Comment (aojea) [reply cluster/gce/config-default.sh:561]: ok, one seems to be GA
```
kg/credentialprovider/plugins.go:      return utilfeature.DefaultFeatureGate.Enabled(features.DisableKubeletCloudCredentialProviders)
pkg/features/kube_features.go:  DisableKubeletCloudCredentialProviders featuregate.Feature = "DisableKubeletCloudCredentialProviders"
pkg/features/kube_features.go:  DisableKubeletCloudCredentialProviders: {Default: false, PreRelease: featuregate.Alpha},
aojea@aojea:~/src/kubernetes$ grep -r KubeletCredentialProviders pkg/
pkg/features/kube_features.go:  KubeletCredentialProviders featuregate.Feature = "KubeletCredentialProviders"
pkg/features/kube_features.go:  KubeletCredentialProviders: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.28
```

**Comment by aojea** [comment]
ok, this is passing now and running the cloud-controller-manager

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/114557/pull-kubernetes-e2e-gce/1651321873713074176/artifacts/e2e-f638ea44c7-674b9-master/cloud-controller-manager.log

```
I0426 21:11:24.992565      10 gen.go:25575] GCERoutes.List(context.TODO.WithCancel.WithDeadline(2023-04-26 22:11:24.583345404 +0000 UTC m=+4653.227494780 [59m59.590769032s]), ..., (name eq e2e-f638ea44c7-674b9-.*) (network eq https://www.googleapis.com/compute/v1/projects/k8s-infra-e2e-boskos-031/global/networks/e2e-f638ea44c7-674b9) (description eq k8s-node-route)) = [4 items], <nil>
I0426 21:11:24.992643      10 route_controller.go:216] action for Node "e2e-f638ea44c7-674b9-master" with CIDR "10.64.3.0/24": "keep"
I0426 21:11:24.992657      10 route_controller.go:216] action for Node "e2e-f638ea44c7-674b9-minion-group-9k00" with CIDR "10.64.2.0/24": "keep"
I0426 21:11:24.992666      10 route_controller.go:216] action for Node "e2e-f638ea44c7-674b9-minion-group-ffwv" with CIDR "10.64.1.0/24": "keep"
I0426 21:11:24.992675      10 route_controller.go:216] action for Node "e2e-f638ea44c7-674b9-minion-group-z4rs" with CIDR "10.64.0.0/24": "keep"
```
@cheftako @andrewsykim should we have the route_controller running?

**Comment by dims** [comment]
@aojea please rebase this PR, then i'll land it quickly!

**Comment by moficodes** [comment]
@aojea 

1.28 Bug Triage Shadow here.

Just checking-in, if this is still on track for the 1.28 release?

**Comment by furkatgofurov7** [comment]
@aojea hey, Bug Triage Lead here 👋🏼 

Can we please rebase this PR?

FYI, the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about less than 2 weeks from now).

**Comment by moficodes** [comment]
Week 9 update: Code Freeze is approaching!

Hey! We are currently in Week 9, and Code Freeze is just over a week away. It will begin at 01:00 UTC on Wednesday, 19th July 2023 / 18:00 PDT on Tuesday, 18th July 2023. Please ensure all necessary submissions are made before the deadline. Let me know if you need any assistance. :)

**Comment by elmiko** [comment]
@aojea curious if you have time to rebase this or if we (the sig) might take up the PR?

**Comment by aojea** [comment]
@elmiko feel free to clone the PR and take over, I think that is a good idea , I'm happy to assist and let you drive this, so I do not become a blocker, I

**Comment by elmiko** [comment]
@aojea ack, thanks for the update. i will most likely clone this PR on monday, give it some local tests, and the continue to drive it forward. i will leave a message here when i've got something to share.

**Comment by elmiko** [comment]
i have created https://github.com/kubernetes/kubernetes/pull/120263 as the continuation of this work. unless there are objections, i will close this issue and we can continue the work on #120263 .

/close

**Pull request closed without merging**